vispy/visuals/tube.py

```diff
@@ -60,63 +60,122 @@ def __init__(self, points, radius=1.0,
-
-        segments = len(points) - 1
-
         # if single radius, convert to list of radii
         if not isinstance(radius, collections.abc.Iterable):
             radius = [radius] * len(points)
         elif len(radius) != len(points):
             raise ValueError('Length of radii list must match points.')
 
         # get the positions of each vertex
-        grid = np.zeros((len(points), tube_points, 3))
-        for i in range(len(points)):
-            pos = points[i]
-            normal = normals[i]
-            binormal = binormals[i]
-            r = radius[i]
-
-            # Add a vertex for each point on the circle
-            v = np.arange(tube_points,
-                          dtype=np.float32) / tube_points * 2 * np.pi
-            cx = -1. * r * np.cos(v)
-            cy = r * np.sin(v)
-            grid[i] = (pos + cx[:, np.newaxis]*normal +
-                       cy[:, np.newaxis]*binormal)
+        vertices = _get_vertices(
+            points=points,
+            closed=closed,
+            tube_points=tube_points,
+            radius=radius)
 
         # construct the mesh
-        indices = []
-        for i in range(segments):
-            for j in range(tube_points):
-                ip = (i+1) % segments if closed else i+1
-                jp = (j+1) % tube_points
-
-                index_a = i*tube_points + j
-                index_b = ip*tube_points + j
-                index_c = ip*tube_points + jp
-                index_d = i*tube_points + jp
-
-                indices.append([index_a, index_b, index_d])
-                indices.append([index_b, index_c, index_d])
-
-        vertices = grid.reshape(grid.shape[0]*grid.shape[1], 3)
+        indices = _get_indices(points=points, closed=closed,
+                              tube_points=tube_points)
 
         color = ColorArray(color)
         if vertex_colors is None:
             point_colors = np.resize(color.rgba,
                                      (len(points), 4))
             vertex_colors = np.repeat(point_colors, tube_points, axis=0)
 
-        indices = np.array(indices, dtype=np.uint32)
-
         MeshVisual.__init__(self, vertices, indices,
                             vertex_colors=vertex_colors,
                             face_colors=face_colors,
                             shading=shading,
                             mode=mode)
 
+    def set_data(self, points, radius=1.0,
+                 closed=False, color='purple',
+                 tube_points=8, shading='smooth',
+                 vertex_colors=None, face_colors=None,
+                 mode='triangles'):
+
+        points = np.array(points).astype(float)
+        # if single radius, convert to list of radii
+        if not isinstance(radius, collections.abc.Iterable):
+            radius = [radius] * len(points)
+        elif len(radius) != len(points):
+            raise ValueError('Length of radii list must match points.')
```

Comment:
This is repeated from the code in `__init__`. Can this be put in a helper method and called from both locations?

Actually I see that all of this method is repeated in `__init__`. Maybe create a method that gets `vertices`, `indices`, `vertex_colors`, and `face_colors`, and use it in both? Or is it possible to create an empty MeshVisual and then call this `set_data` from the `__init__` method?